### ISSUE TITLE
dandi upload parallel jobs

### ISSUE THREAD

**Issue by bendichter** [dandi/dandi-cli#927]
When you do `dandi upload`, it's not clear how many parallel jobs are used by default. I would have expected the program to automatically detect how many jobs are available. Upon some digging, I found this `iter_upload` method, which claims the number of jobs defaults to 5 but does not set it to 5 in the method signature. 

https://github.com/dandi/dandi-cli/blob/a1a7fd553cf300967adc060302c7e6f6c1b921c8/dandi/files.py#L765-L792

Then the default to 5 is set down here:

https://github.com/dandi/dandi-cli/blob/a1a7fd553cf300967adc060302c7e6f6c1b921c8/dandi/files.py#L845

If you really want to hard-code 5, wouldn't it be easier to just do it in the function signature like:

```python
jobs: Optional[int] = 5
```
And then it would be nice to indicate this in the cli help.

A better option might be to use what is [now default for Python >=3.8](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor): `max_workers=min(32, os.cpu_count() + 4)`

**Comment by jwodder** [comment]
@bendichter 

> I would have expected the program to automatically detect how many jobs are available.

Detect what, exactly?  CPUs?  Cores?

> If you really want to hard-code 5, wouldn't it be easier to just do it in the function signature like:

No, doing it that way means that passing `None` would no longer default to 5.  `jobs=None` has to be handled inside the function body.

> A better option might be to use what is [now default for Python >=3.8](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor): max_workers=min(32, os.cpu_count() + 4)

Note that `iter_upload()` manages the upload of a single asset, and its `jobs` parameter is the number of concurrent upload threads for just that asset.  When uploading, there will be multiple assets uploaded in parallel, each with their own collection of threads, so setting the per-asset thread count to the "max" would mean multiple times that many threads.

**Comment by bendichter** [comment]
@jwodder how many jobs are spun up for all across assets?

**Comment by jwodder** [comment]
@bendichter Well, [pyout is told to limit the number of assets uploaded at once](https://github.com/dandi/dandi-cli/blob/a1a7fd553cf300967adc060302c7e6f6c1b921c8/dandi/upload.py#L283-L285) to a value specified on the command line, defaulting to `concurrent.futures`'s current `max_workers` default, and then [dandi-cli itself also limits the number of concurrent assets to 10](https://github.com/dandi/dandi-cli/blob/a1a7fd553cf300967adc060302c7e6f6c1b921c8/dandi/upload.py#L289), and each asset upload gets one main thread and (by default) 5 upload threads.

**Comment by bendichter** [comment]
OK so it looks like you do already use the logic I was suggesting for determining the number of parallel jobs. Then with a cap of 10? Why? Would it make sense to set the default to `max_workers=min(10, os.cpu_count() + 4)`?
It would be nice to have some indication of this in the cli help. As a user I would expect if no default behavior is described that only a single worker is used by default, which is clearly not the case.

**Comment by yarikoptic** [comment]
well, I guess I love magic too much.  That is why I like when tools do not need more options to be specified than necessary while performing efficiently. That is why we do parallelize by default and you can consider it to be an "opt-out" feature to go back to a single worker.

**Comment by bendichter** [comment]
@yarikoptic yes, there is nothing wrong with parallelizing by default! I just want that do be documented.

**Comment by yarikoptic** [comment]
> @yarikoptic yes, there is nothing wrong with parallelizing by default! I just want that do be documented.

agree with that. @jwodder - let's figure out how to achieve that without heavy docs duplication etc.

**Comment by jwodder** [comment]
@yarikoptic So just document the default behavior of `dandi upload --jobs`?  To be honest, [as I described it above](https://github.com/dandi/dandi-cli/issues/927#issuecomment-1052797761), it doesn't make much sense; should the dandi-imposed limit of 10 be removed as well?

**Comment by yarikoptic** [comment]
d'oh -- I have read `min` as `max` ;-) but the issue stands -- deciding on # of parallel uploads (IO) based on number of CPUs makes little sense since we are trying to avoid wasting on waiting for IO. Thus any dynamic settings for `None` probably makes not much sense.  Our goal should be to obtain some sensible performance by default. Uploading 5 files in parallel by default with 5 threads within each makes sense to me. Ideally of cause we should do some smarter "sensing" and scaling of number of jobs/threads based on file sizes and observed performance, and then document just that "By default  dandi-cli scales number of jobs automagically". While we do not have magic here -- let's "hardcode" to some sensible setting and just leave it for use to tune up. There will be no `min` or `max` on anything. And that is what I meant. WDYT?

**Comment by jwodder** [comment]
@yarikoptic So just have `--jobs` default to a fixed number instead of using a default based on CPU count?  Sure.  What should that number be?

**Comment by yarikoptic** [comment]
`5:5` -- 5 jobs with 5 threads per job
